torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2346,13 +2347,16 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
                 grad = param._saved_grad_shard  # type: ignore[attr-defined]
             else:
                 # Currently the only way for _is_sharded to be False is if
-                # world_size == 1. This could be relaxed in the future, e.g,
-                # no sharding like PyTorch DDP, in which case grads should be
-                # all-reduced here.
+                # world_size == 1 or sharding_strategy is NO_SHARD.
                 assert (
-                    self.world_size == 1
+                    self.world_size == 1 or self.sharding_strategy == ShardingStrategy.NO_SHARD
                 ), "Currently the only way for _is_sharded to be False is \
-                    world_size == 1"
+                    world_size == 1 or sharding_stratagy is set to be NO_SHARD"
+                if self.sharding_strategy == ShardingStrategy.NO_SHARD:
+                    dist.all_reduce(param.grad, group=self.process_group)
```

Comment:
this will be a synchronous comm op? I assume there will be followup PRs to make this async?

Reply:
it will sync on post_backward stream, post_backward stream will be in parallel with default computation stream, so it is like async all reduce. in the end of backward pass, it waits for post_backward stream instead of waiting for future object of async all reduce

Reply:
this is the same as how reduce_scatter and all_gather are called here. Also checked trace and computation&communication is overlapped